10_Examples/jett-onboarding/src/publisher_member_function.cpp

```diff
@@ -0,0 +1,56 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <memory>
+
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs/msg/string.hpp"
+
+using namespace std::chrono_literals;
+
+/* This example creates a subclass of Node and uses std::bind() to register a
+ * member function as a callback from the timer. */
+
+class MinimalPublisher : public rclcpp::Node
+{
+public:
+  MinimalPublisher()
+      : Node("minimal_publisher"), count_(0)
+  {
+    publisher_ = this->create_publisher<std_msgs::msg::String>("topic", 10);
+    timer_ = this->create_wall_timer(
+        500ms, std::bind(&MinimalPublisher::timer_callback, this));
+  }
+
+private:
+  void timer_callback()
+  {
+    auto message = std_msgs::msg::String();
+    message.data = "Hello, world! " + std::to_string(count_++);
+    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
+    publisher_->publish(message);
+  }
+  rclcpp::TimerBase::SharedPtr timer_;
+  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
+  size_t count_;
+};
+
+int main(int argc, char *argv[])
+{
+  rclcpp::init(argc, argv);
+  rclcpp::spin(std::make_shared<MinimalPublisher>());
```

Comment:
Wont go into too much depth because the ROS tutorials are pretty good, but I do want to mention a bit about spin.

The spin function is basically this:
```
while(true){
if there are any callbacks waiting in line/queue, process them in order they arrived.
}
```

The important things from that example being that:
- Spin is blocking (i.e. once you call this line, it wont continue until ROS dies / you kill the program)
- Callbacks are NOT processed in parallel. So if you have a callback that waits a really long time, all the other callbacks are stuck in line and that data wont get processed. In general, callbacks should be short and we want to grab the incoming data and move it somewhere else for processing so we can be free to handle more callbacks.

Reply:
In this case, we arent subscribing to any ROS topics, but the time processes callback functions in the same way.